fastfield return iterator instead fill vec

PSeitz opened the pull request:
return iterator from get_vals method. This will allow to save on unnecessary vec allocations.

fulmicoton commented on fastfield_codecs/src/blockwise_linear.rs:
```diff
@@ -300,13 +300,9 @@ impl FastFieldCodec for BlockwiseLinearCodec {
-        {
-            return None;
-        }
+            .checked_add(theorethical_maximum_offset)?;
```
I prefer the earlier version... By far actually.

In the new version the intent is hidden.

PSeitz replied:
it's from a clippt lint, I also think the previous version is more clear. We could use [overflowing_add](https://doc.rust-lang.org/std/primitive.i8.html#method.overflowing_add) instead and check the bool, or disable the lint

fulmicoton replied:
What about disabling the lint locally?

fulmicoton commented on fastfield_codecs/src/column.rs:
```diff
@@ -8,23 +10,14 @@ pub trait Column<T = u64> {
     /// # Panics
     ///
-    /// May panic if `start + output.len()` is greater than
+    /// May panic if `range.end()` is greater than
```
```suggestion
    /// May panic if `range.end` is greater than
```

fulmicoton commented on src/fastfield/facet_reader.rs:
```diff
@@ -76,7 +76,8 @@ impl FacetReader {
 
     /// Return the list of facet ordinals associated to a document.
     pub fn facet_ords(&self, doc: DocId, output: &mut Vec<u64>) {
-        self.term_ords.get_vals(doc, output);
```
so here, this is a net loss.

fulmicoton commented on fastfield_codecs/src/linear.rs:
```diff
@@ -204,13 +204,9 @@ impl FastFieldCodec for LinearCodec {
-        {
-            return None;
-        }
+            .checked_add(theorethical_maximum_offset)?;
```
same as before. I prefer the earlier version.

fulmicoton commented:
Do you have benchmark to confirm we don't have a regression in search?

The `fill` to buffer indirection point is that it get rids of dynamic dispatch.

PSeitz commented:
> Do you have benchmark to confirm we don't have a regression in search?
> 
> The `fill` to buffer indirection point is that it get rids of dynamic dispatch.

I removed the dynamic dispatch and added a bench. It is still slower, which I don't understand yet. 

```
Vec version - Most time spent in MultiValuedFastFieldReader<Item>::get_vals
running 1 test
test fastfield::bench::bench_multi_value_fflookup                             ... bench:     758,148 ns/iter (+/- 13,857)

+   93,36%    15,12%  tantivy-cbbeafa  tantivy-cbbeafab3d8c2ac9      [.] test::bench::Bencher::iter                                                                ▒
+   78,22%    78,20%  tantivy-cbbeafa  tantivy-cbbeafab3d8c2ac9      [.] tantivy::fastfield::multivalued::reader::MultiValuedFastFieldReader<Item>::get_vals       ▒


Iter version Static dispatch - Most time spent in sum
running 1 test
test fastfield::bench::bench_multi_value_fflookup                             ... bench:   1,382,918 ns/iter (+/- 32,6

+   96,21%    32,31%  tantivy-ea620f5  tantivy-ea620f5d7d4bf2df      [.] test::bench::Bencher::iter                                                               ▒
+   50,75%    50,73%  tantivy-ea620f5  tantivy-ea620f5d7d4bf2df      [.] <u64 as core::iter::traits::accum::Sum>::sum                                             ▒
+   13,12%    13,10%  tantivy-ea620f5  tantivy-ea620f5d7d4bf2df      [.] tantivy::fastfield::multivalued::reader::MultiValuedFastFieldReader<Item>::get_vals      ▒



Iter version dynamic dispatch
running 1 test
test fastfield::bench::bench_multi_value_fflookup                             ... bench:   1,550,563 ns/iter (+/- 31,509

```

fulmicoton commented:
@PSeitz Let's stick to the Vec filling solution then. It is much easier to reason with and allocation is usually not an issue (we reuse the same buffer over and over).
